specs/agents/tracing-distributed-tracing.md

```diff
@@ -191,6 +191,65 @@ Implementation note: The `traceparent` and `tracestate` specifications only allo
 This implementation guarantees backwards compatibility with our older agents and compatibility with current OpenTelemetry instrumentations.
 We will eventually remove the `elasticapmtraceparent` entirely as soon as we can safely assume that most users have already upgraded all their agents.
 
+### Baggage
+
+To propagate distributed context, we implement the [W3C Baggage](https://www.w3.org/TR/baggage/) specification.
+
+Agents MUST parse, validate, and attach the `baggage` header according [to the W3C specification](https://www.w3.org/TR/baggage/).
+
+In addition to the W3C specification, agents also SHOULD offer users 2 ways to use the values propagated via baggage:
+- Offer a Baggage API to manually manipulate and read baggage values - this is preferable the OpenTelemetry API.
+- Offer baggage related configuration to automatically store baggage values on events.
+
+#### Baggage API
+
+Agents SHOULD integrate with the existing OpenTelemetry API. Users should be able to use the OpenTelemetry API to interact with the baggage maintained by the agent. The primary way to interact with baggage SHOULD be the OpenTelemetry API.
+
+In case using the OpenTelemetry API isn't feasible in a given language, the agent SHOULD extend the proprietary agent API to interact with baggage.
+
+This API MUST offer the following functionalities:
+- Adding a new baggage item
+- Changing an existing baggage item
+- Removing an existing baggage item
+- Reading a specific baggage item
+- Reading all baggage items
+
+#### Baggage related configurations
+
+The following configurations enable users to automatically store baggage items on a given event without any code change. Baggage items with matching keys are stored in `otel.attributes`, except on errors.
+
+`baggage_to_attach_on_transactions` configuration
```

Comment:
Do we want to expose this implementation detail to our users? 

Can we simplify this to just `baggage_to_attach` and let the server decide if it wants to store on transaction/span/error?

Reply:
I think with the current default this is fine - currently we always attach baggage, so users don't really have to care about these configs, unless they want to fine-tune it, and from that point, I think having 3 configs is beneficial.

> Can we simplify this to just `baggage_to_attach` and let the server decide if it wants to store on transaction/span/error?

Maybe I don't see something, but from my understanding that's not feasible. Baggage is a mechanism to propagate the data with distributed traces, but by default it's not stored anywhere. E.g. the [W3C spec](https://www.w3.org/TR/baggage/) only defines how to propagate it (so passing it from one service to another), but storing is not part of it. So unless we store the values somewhere, they won't get to APM Server.

Of course baggage is stored, when an agent (via API or via the configs we propose here) decides to capture it, but that decision is done by the agent. If the agent does not store/send it, the APM Server won't have the baggage values. So letting the server decide would mean that agents would need to store and send all baggage all the time, so that the server can make the decision, which isn't very good.

So overall I feel this should be done in the agent and not in the server.

Reply:
I don't fully understand how and what the server would decide, but you have a very good point regarding 
> why would a user have to care about `I dont want baggage on errors` vs `I only want these specific baggage keys stored`

So yeah, having just 1 config (instead of 3) on the agent side is worth considering. But I still don't think the server should be involved in this feature, other than just storing what it receives. 

So what I could imagine is this:
We only have `baggage_to_attach` as a single config in the agent. Then it'll attach baggage items with matching keys on every event (meaning on transaction, but on spans and errors as well), and then the server will just store them as is.

Reply:
> why would a user have to care about I dont want baggage on errors vs I only want these specific baggage keys stored.

I think we don't know, if they would care or not. One aspect is storage cost for example. storing a baggage item on a transaction vs. on all spans can be a difference in magnitudes regarding the storage impact. 

I agree that we should make it as simple as possible out of the box (so that users would not need to care if they don't want to), but, provide a possibility for configuring in the case they care.  

The baggage as a concept is primarily a vehicle for customization. What I mean by that is that (apart from some exceptions) the primary use case for baggage will be that users explicitly write some baggage (with their custom data and their specific use case in mind) and read it on some downstream services.
So it's not something that they would expect to happen as an "out of the box feature", they will need to write the baggage _explicitly_. (Again that's maybe an argument why baggage rather should be opt-in, vs. what we came up before, not sure, just thinking out loud).

My point is: If users do something explicitly (i.e. customize their data) they would like it to be:

1. easy to use and configure
2. but also, they would like to have as much control on the feature as possible

We don't know what they will use the baggage for. In particular we don't know:

- do they need the baggage on all signals / data (i.e. transactions, spans, errors)?
- what is the potential impact of baggage on the storage costs if it's being attached to _every_ signal (AND we don't provide a way to restrict that to only certain type of data)?
- do they care about that potential storage impact?

I would assume, the above will differ from use case to use case. If we come up with an opinionated solution (in either direction) there will be users who will ask for the opposite. 

That's why I think that especially on this feature (that is meant for data customization purposes) the possibility of doing fine-grained configuration is important.

Reply:
I suggest let's try to come to a conclusion on this.

All points added here are valid and I agree with those (so won't repeat :) )

Given that currently the suggestion is still have `*` as default, I think having 3 configs is fine. The main thinking behind this is the following: by default we attach everything, so users don't need to touch any of these configs, unless they want to fine-tune and if they want to fine-tune, then let's enable them with appropriate configs.

So, I'd suggest let's keep it as is for now (meaning 3 configs). I personally don't have a super string opinion on either and we can still make changes later if we get user feedback. Also keep in mind we only have this in 2 agents at the moment, so we are in a good place to still may make changes once we get feedback.

@Mpdreamz does that sound ok? Otherwise, we can discuss in the weekly meeting as well.

Reply:
> But I still don't think the server should be involved in this feature, other than just storing what it receives.

I actually think the reverse is easier for users. Agents `capture+send` and server `stores`. 

Pushing these specialized `storage` configs to agents makes it exponentially harder for users to push down to N services over N machines vs setting storage related configs at the central receiving end. 

> I suggest let's try to come to a conclusion on this. 

I think I made my points, happy to concede and move on in the interest of shipping 🚢 😄

Reply:
Current suggestion is a single agent side config - that's what we arrived to in the agents weekly: https://github.com/elastic/apm/pull/820#issuecomment-1671024912